server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -513,8 +537,94 @@ public ClusterMetadataManifest getLatestClusterMetadataManifest(String clusterNa
+            throw new IllegalStateException("Error while fetching previous UUIDs from remote store");
+        }
+    }
+
+    private Set<String> getAllClusterUUIDs(String clusterName) throws IOException {
+        Map<String, BlobContainer> clusterUUIDMetadata = clusterUUIDContainer(clusterName).children();
+        if (clusterUUIDMetadata == null) {
+            return Collections.emptySet();
+        }
+        return Collections.unmodifiableSet(clusterUUIDMetadata.keySet());
+    }
+
+    private Map<String, ClusterMetadataManifest> getLatestManifestForAllClusterUUIDs(String clusterName, Set<String> clusterUUIDs) {
+        Map<String, ClusterMetadataManifest> manifestsByClusterUUID = new HashMap<>();
+        for (String clusterUUID : clusterUUIDs) {
+            try {
+                ClusterMetadataManifest manifest = getLatestClusterMetadataManifest(clusterName, clusterUUID);
+                manifestsByClusterUUID.put(clusterUUID, manifest);
+            } catch (Exception e) {
+                logger.error("Exception in fetching manifest for clusterUUID: {}", clusterUUID);
+                // Throwing the exception since we may not be able to construct the clusterUUID chain if some manifest file is missing.
+                throw new IllegalStateException(" Error while fetching manifest data from remote store");
+            }
+        }
+        return manifestsByClusterUUID;
+    }
+
+    /**
+     * This method creates a valid cluster UUID chain.
+     *
+     * @param manifestsByClusterUUID Map of latest ClusterMetadataManifest for every cluster UUID
+     * @return List of cluster UUIDs. The first element is the most recent cluster UUID in the chain
+     */
+    private List<String> createClusterChain(final Map<String, ClusterMetadataManifest> manifestsByClusterUUID) {
+        final Map<String, String> clusterUUIDGraph = manifestsByClusterUUID.values()
+            .stream()
+            .collect(Collectors.toMap(ClusterMetadataManifest::getClusterUUID, ClusterMetadataManifest::getPreviousClusterUUID));
+        final List<String> validClusterUUIDs = manifestsByClusterUUID.values()
+            .stream()
+            .filter(m -> !isInvalidClusterUUID(m) && !clusterUUIDGraph.containsValue(m.getClusterUUID()))
+            .map(ClusterMetadataManifest::getClusterUUID)
+            .collect(Collectors.toList());
+        if (validClusterUUIDs.isEmpty()) {
+            logger.info("There is no valid previous cluster UUID");
+            return Collections.emptyList();
+        }
+        if (validClusterUUIDs.size() > 1) {
+            logger.error(
+                "There are more than 1 valid top level previous cluster UUIDs which are part of different chains. Cluster UUIDs: {} . "
+                    + "Cleanup the invalid cluster UUID data",
+                validClusterUUIDs
+            );
+            throw new IllegalStateException("More than 1 valid top level cluster UUIDs present in remote store");
```

Comment:
How do we cleanup in such cases?

Reply:
Normally this should not happen. It could happen in exceptional scenarios such as:
-  2 clusters with the same name are publishing to the same repo.
- Customer cleaned up partial data from the repo manually.

In this case, one way to delete the data for all UUID except one valid UUID. But in case this is difficult to identify, user would have to flush all the data.

Reply:
we need to be more explicit in the error. Something like -> The system has ended into mutiple valid cluster states in the remote. Please check their latest manifest to decide which one you want to keep. 

In documentation we can explain, when this can happen and how to fix.